src/main/java/com/google/devtools/build/lib/sandbox/SandboxStash.java

```diff
@@ -242,4 +262,33 @@ static void clean(TreeDeleter treeDeleter, Path outputBase) {
       logger.atWarning().withCause(e).log("Failed to clean sandbox stash %s", stashDir);
     }
   }
+
+  /**
+   * Test actions are guaranteed to have a runfiles directory with the test name as part of the
+   * name. The path to the directory is unique between tests. If two tests (foo and bar) have the
+   * directory <source-root>/pkg/my_runfiles as part of their runfiles and this directory contains
+   * 1000 files, we would be symlinking the 1000 files for each test since the paths do not
+   * coincide. To make sure we can reuse the runfiles directory we must rename the old runfiles
+   * directory for the action that was stashed to the path that is expected by the current test.
+   */
+  private static boolean isTestAction(String mnemonic) {
+    return mnemonic.equals(TEST_RUNNER_MNEMONIC);
+  }
+
+  /**
+   * Test actions are split in two spawns. The first one runs the test and the second generates the
+   * XML output from the test log. We do not want the second spawn to reuse the stash because it
+   * doesn't contain the inputs needed to run the test; if it reused it, it would be expensive in
+   * two ways: it would have to clean up all the inputs, and it would destroy a valid stash that a
+   * different test could potentially use.
+   *
+   * We identify the second spawn by having a single output.
+   */
+  private static boolean isTestXmlGenerationAction(String mnemonic, SandboxOutputs outputs) {
```

Comment:
Just something I thought of while going over this PR: Did you check whether the separate coverage spawn used with split post-processing also requires special handling?

Reply:
This condition is meant to catch that. Is there a third spawn?

Reply:
Based on the method name, this is handling the one created by `StandaloneTestStrategy#createXmlGeneratingSpawn`. But there is also the one created by `StandaloneTestStrategy#createCoveragePostProcessingSpawn`.

Reply:
Ok, good thing you mentioned it, so it works but by pure chance because it also has one file as output. I will rename the method. Thanks!